File: app/services/cache.py
"""
Async Redis cache service.
Provides caching of LLM responses keyed by the full request shape (messages,
model, project, and generation parameters) so two requests that differ only
in temperature/top_p/etc. never collide.
"""

import hashlib
import json
import logging
from typing import Optional

from app.config import get_settings
from app.database.redis import get_redis

logger = logging.getLogger(__name__)

# Cache key prefix
CACHE_PREFIX = "gw:cache:"
CACHE_STATS_KEY = "gw:cache:stats"
PROJECT_INDEX_PREFIX = "gw:cache:project:"


def _build_cache_key(
    messages: list[dict],
    model: str,
    project: str = "default",
    params: Optional[dict] = None,
) -> str:
    """
    Build a deterministic cache key from messages, model, project, and the
    generation parameters that affect the response content.
    """
    effective_params = {
        k: v for k, v in (params or {}).items()
        if v is not None and k in {
            "temperature", "top_p", "max_tokens", "stop",
            "frequency_penalty", "presence_penalty", "n",
            "tools", "tool_choice", "response_format",
        }
    }
    normalized = json.dumps(
        {
            "messages": messages,
            "model": model,
            "project": project,
            "params": effective_params,
        },
        sort_keys=True,
        ensure_ascii=True,
        default=str,
    )
    hash_digest = hashlib.sha256(normalized.encode("utf-8")).hexdigest()
    return f"{CACHE_PREFIX}{hash_digest}"


async def get_cached_response(
    messages: list[dict],
    model: str,
    project: str = "default",
    params: Optional[dict] = None,
) -> Optional[dict]:
    """
    Look up a cached LLM response.

    Returns:
        Cached response dict if found, None if cache miss.
    """
    try:
        redis = await get_redis()
        key = _build_cache_key(messages, model, project, params)
        cached = await redis.get(key)

        if cached:
            logger.debug(f"Cache HIT: {key[:20]}...")
            await redis.hincrby(CACHE_STATS_KEY, "hits", 1)
            return json.loads(cached)

        logger.debug(f"Cache MISS: {key[:20]}...")
        await redis.hincrby(CACHE_STATS_KEY, "misses", 1)
        return None

    except Exception as e:
        logger.error(f"Cache read error: {e}")
        return None


async def set_cached_response(
    messages: list[dict],
    model: str,
    response: dict,
    project: str = "default",
    ttl: Optional[int] = None,
    params: Optional[dict] = None,
):
    """
    Store an LLM response in cache.

    Args:
        messages: The conversation messages
        model: Model identifier
        response: The LLM response dict to cache
        project: Project name
        ttl: Time-to-live in seconds (default from settings)
        params: Generation parameters that affect the response
    """
    try:
        settings = get_settings()
        redis = await get_redis()
        key = _build_cache_key(messages, model, project, params)
        expire = ttl or settings.cache_ttl

        await redis.set(
            key,
            json.dumps(response, ensure_ascii=False),
            ex=expire,
        )
        # Maintain the project index so per-project invalidation works
        index_key = f"{PROJECT_INDEX_PREFIX}{project}"
        await redis.sadd(index_key, key)
        await redis.expire(index_key, expire)

        logger.debug(f"Cache SET: {key[:20]}... (ttl={expire}s)")

    except Exception as e:
        logger.error(f"Cache write error: {e}")


async def invalidate_project_cache(project: str) -> int:
    """
    Invalidate all cached responses for a specific project.
    Uses the project key index maintained by set_cached_response.

    Returns:
        Number of cache entries removed.
    """
    try:
        redis = await get_redis()
        index_key = f"{PROJECT_INDEX_PREFIX}{project}"
        keys = await redis.smembers(index_key)

        removed = 0
        if keys:
            removed = await redis.delete(*keys)
        await redis.delete(index_key)
        logger.info(f"Invalidated {removed} cached entries for project '{project}'")
        return removed

    except Exception as e:
        logger.error(f"Cache invalidation error: {e}")
        return 0


async def get_cache_stats() -> dict:
    """Get cache hit/miss statistics."""
    try:
        redis = await get_redis()
        stats = await redis.hgetall(CACHE_STATS_KEY)
        # redis-py returns bytes keys/values on raw hashes
        decoded = {
            (k.decode() if isinstance(k, bytes) else k):
                (v.decode() if isinstance(v, bytes) else v)
            for k, v in stats.items()
        }
        return {
            "hits": int(decoded.get("hits", 0)),
            "misses": int(decoded.get("misses", 0)),
        }
    except Exception as e:
        logger.error(f"Cache stats error: {e}")
        return {"hits": 0, "misses": 0}

File: gateway/internal/cache/cache.go
// Package cache is the Redis-backed response cache.
//
// Key derivation is param-aware: two requests differing only in e.g.
// temperature never share a cached answer. Project invalidation relies on a
// per-project set index (gw:cache:project:{name}).
package cache

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"log/slog"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"
)

const (
	cachePrefix    = "gw:cache:"
	projectIndexPfx = "gw:cache:project:"
	statsKey        = "gw:cache:stats"
)

// Store is the public cache interface.
type Store struct {
	rdb redis.UniversalClient
	ttl time.Duration
}

// New builds a cache bound to the given Redis client.
func New(rdb redis.UniversalClient, ttlSecs int) *Store {
	if ttlSecs <= 0 {
		ttlSecs = 86400
	}
	return &Store{rdb: rdb, ttl: time.Duration(ttlSecs) * time.Second}
}

// paramFields is the whitelist of params that participate in the cache key.
var paramFields = []string{
	"temperature", "top_p", "max_tokens", "stop",
	"frequency_penalty", "presence_penalty", "n",
	"tools", "tool_choice", "response_format",
}

// Key builds a deterministic cache key for one request.
func Key(messages any, model, project string, params map[string]any) string {
	effective := map[string]any{}
	for _, k := range paramFields {
		if v, ok := params[k]; ok && v != nil {
			effective[k] = v
		}
	}
	blob := map[string]any{
		"messages": messages,
		"model":    model,
		"project":  project,
		"params":   effective,
	}
	buf, _ := json.Marshal(blob)
	sum := sha256.Sum256(buf)
	return cachePrefix + hex.EncodeToString(sum[:])
}

// Get retrieves a cached response. Hits bump the stats counter.
func (s *Store) Get(ctx context.Context, key string) (map[string]any, bool) {
	raw, err := s.rdb.Get(ctx, key).Result()
	if errors.Is(err, redis.Nil) {
		s.incr(ctx, "misses")
		return nil, false
	}
	if err != nil {
		slog.Warn("cache read failed", "err", err)
		return nil, false
	}
	s.incr(ctx, "hits")
	out := map[string]any{}
	if err := json.Unmarshal([]byte(raw), &out); err != nil {
		return nil, false
	}
	return out, true
}

// Set stores a response; also appends the key to the project's index set.
func (s *Store) Set(ctx context.Context, key string, response map[string]any, project string) {
	buf, err := json.Marshal(response)
	if err != nil {
		return
	}
	pipe := s.rdb.TxPipeline()
	pipe.Set(ctx, key, buf, s.ttl)
	idxKey := projectIndexPfx + project
	pipe.SAdd(ctx, idxKey, key)
	pipe.Expire(ctx, idxKey, s.ttl)
	if _, err := pipe.Exec(ctx); err != nil {
		slog.Warn("cache write failed", "err", err)
	}
}

// InvalidateProject drops every cached entry for a project.
func (s *Store) InvalidateProject(ctx context.Context, project string) int {
	idxKey := projectIndexPfx + project
	keys, err := s.rdb.SMembers(ctx, idxKey).Result()
	if err != nil || len(keys) == 0 {
		_ = s.rdb.Del(ctx, idxKey).Err()
		return 0
	}
	n := 0
	if keys != nil {
		dn, err := s.rdb.Del(ctx, keys...).Result()
		if err == nil {
			n = int(dn)
		}
	}
	s.rdb.Del(ctx, idxKey)
	return n
}

// Stats returns the hit/miss counters.
func (s *Store) Stats(ctx context.Context) map[string]int {
	out := map[string]int{"hits": 0, "misses": 0}
	vals, err := s.rdb.HGetAll(ctx, statsKey).Result()
	if err != nil {
		return out
	}
	if h, ok := vals["hits"]; ok {
		if n, e := strconv.Atoi(h); e == nil {
			out["hits"] = n
		}
	}
	if m, ok := vals["misses"]; ok {
		if n, e := strconv.Atoi(m); e == nil {
			out["misses"] = n
		}
	}
	return out
}

func (s *Store) incr(ctx context.Context, field string) {
	if err := s.rdb.HIncrBy(ctx, statsKey, field, 1).Err(); err != nil {
		slog.Debug("stats incr failed", "err", err)
	}
}
